### PR TITLE
Fix OSM on a OSL activated layer

### PR DESCRIPTION
**Pull request by NapOli1084**
## Description

This change follows #19214 by @kosorin, which allowed mod-tap hold action on one shot layers. The purpose of this change is to do the same for one shot mods and allow them on one shot layers.

I attempted adding `OSM(MOD_LSFT)` on one of my `OSL` activated layers in order to use it along with accented letters I have on that layer. I realized it didn't work and after looking into the code I noticed the condition that prevents deactivating one shot layers with mods, which wasn't considering one shot mods, but was considering mod-tap. So it seemed like a simple oversight, doesn't seem intended that OSM wasn't considered. At least I can't think of a reason why it would be supported with other mods but not with OSM.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: quantum/action.c [hunk 1]
```diff
@@ -373,7 +373,8 @@ void process_action(keyrecord_t *record, action_t action) {
     if (is_oneshot_layer_active() && event.pressed &&
         (action.kind.id == ACT_USAGE || !(IS_MODIFIER_KEYCODE(action.key.code)
 #    ifndef NO_ACTION_TAPPING
-                                          || (tap_count == 0 && (action.kind.id == ACT_LMODS_TAP || action.kind.id == ACT_RMODS_TAP))
+                                          || ((action.kind.id == ACT_LMODS_TAP || action.kind.id == ACT_RMODS_TAP) &&
+                                                (action.layer_tap.code <= MODS_TAP_TOGGLE || tap_count == 0))
```
Comment (NapOli1084) [inline quantum/action.c:377]: `tap_count == 0` accounted for mod-tap, but not for OSM. `action.layer_tap.code <= MODS_TAP_TOGGLE` accounts for `MODS_ONESHOT` and `MODS_TAP_TOGGLE`. TBH I don't know how the latter can occur, can't figure a keycode leading to it. But considering it's for mods also I think it makes sense to consider them too.

**Comment by NapOli1084** [comment]
I intended to look into adding a unit test for this, but I can't seem to build them. I'm getting "undefined reference" linker errors on many functions, seems to be the ones with weak linking. I'm on QMK MSYS, don't know if it's a setup issue. I'll try asking around on Discord.

**Comment by NapOli1084** [comment]
> I intended to look into adding a unit test for this, but I can't seem to build them. I'm getting "undefined reference" linker errors on many functions, seems to be the ones with weak linking. I'm on QMK MSYS, don't know if it's a setup issue. I'll try asking around on Discord.

I didn't figure out my linker issue, but gave it a try anyway, relying on the CI tests. Unfortunately means I failed on first attempt, but second succeeded.

**Review by drashna — APPROVED** [review]

**Review by tzarc — APPROVED** [review]

**Comment by bencollerson** [comment]
This change breaks the functionality of putting one shot mods on a layer and using them on the default (or other) layer. 

I raised an issue related to this: https://github.com/qmk/qmk_firmware/issues/22566

**Comment by elshize** [comment]
Same here, I am relying on the old functionality, and had to revert these changes to make it work again.

**Pull request merged**